### PR TITLE
SDF wall-distance as physics input feature — cross-dataset geometry encoding

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

**Signed Distance Function (SDF) / wall-distance as an explicit physics-informed input feature across CFD benchmarks.**

All current models rely on Fourier embeddings of raw coordinates. However, in CFD, the wall-distance (distance to the nearest solid boundary) is a physically fundamental quantity:

1. It governs boundary layer thickness and turbulence intensity (wall-distance appears explicitly in RANS turbulence models like k-ω SST)
2. It encodes geometry proximity in a way that polynomial/Fourier coordinates do not
3. It has been successfully used in Graph Neural Network CFD surrogates (MeshGraphNet variants) as a physics-informed feature

For 2D airfoil cases (TandemFoil, AirfRANS), wall distance can be computed as the minimum distance from each mesh point to the airfoil surface points. For 3D DrivAerML, the same applies to the car surface.

## Implementation

Compute wall-distance as a per-point feature during data loading and append it to the input feature vector:

```python
# During dataset preprocessing / feature construction:
# For each mesh point, compute minimum distance to any boundary/surface point.
# surface_points: [N_surf, 3] — the solid body surface coordinates
# volume_points:  [N_vol, 3]  — all mesh points

import torch

def compute_wall_distance(volume_points, surface_points):
    """
    Compute minimum distance from each volume point to any surface point.
    Returns wall_dist: [N_vol, 1]
    """
    # Chunked to avoid OOM on large meshes
    chunk_size = 4096
    dists = []
    for i in range(0, len(volume_points), chunk_size):
        chunk = volume_points[i:i+chunk_size]  # [chunk, 3]
        d = torch.cdist(chunk.unsqueeze(0), surface_points.unsqueeze(0))[0]  # [chunk, N_surf]
        dists.append(d.min(dim=-1).values)  # [chunk]
    wall_dist = torch.cat(dists, dim=0).unsqueeze(-1)  # [N_vol, 1]
    return wall_dist
```

Then concatenate `wall_dist` to the input features alongside coordinates and Fourier embeddings. Apply optional log-normalization since wall-distance has long-tail distribution: `log(1 + wall_dist / mean_dist)`.

## Runs (3 datasets)

Run these 3 jobs using `--wandb_group sdf-wall-distance-cross-dataset` on all runs.

### Run 1 — TandemFoil

```bash
python train.py \
  --dataset tandemfoil \
  --optimizer lion \
  --lr 1.25e-4 \
  --cosine-t-max 10 \
  --grad-clip 0.5 \
  --weight-decay 1e-2 \
  --model-slices 64 \
  --model-layers 3 \
  --model-hidden-dim 192 \
  --model-heads 3 \
  --enable-fourier \
  --enable-te-coord-frame \
  --enable-cp-panel \
  --enable-cp-panel-tandem-only \
  --asinh-pressure \
  --residual-prediction \
  --enable-pressure-prior-addition \
  --epochs 999 \
  --ema-decay 0.999 \
  --wandb_group sdf-wall-distance-cross-dataset
```

### Run 2 — AirfRANS

```bash
python train.py \
  --dataset airfrans \
  --airfrans-task full \
  --optimizer adamw \
  --lr 6e-4 \
  --cosine-t-max 50 \
  --grad-clip 1.0 \
  --weight-decay 1e-2 \
  --no-use-ema \
  --enable-fourier \
  --model-layers 2 \
  --model-hidden-dim 256 \
  --model-heads 4 \
  --epochs 999 \
  --wandb_group sdf-wall-distance-cross-dataset
```

### Run 3 — DrivAerML

```bash
SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml \
  --optimizer adamw \
  --lr 5e-4 \
  --cosine-t-max 30 \
  --no-use-ema \
  --enable-fourier \
  --model-layers 4 \
  --model-hidden-dim 512 \
  --model-heads 8 \
  --epochs 999 \
  --batch-size 1 \
  --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 \
  --max-eval-batches 200 \
  --wandb_group sdf-wall-distance-cross-dataset
```

## Current Baselines

| Dataset | Metric | Current Best | Best PR |
|---------|--------|-------------|---------|
| TandemFoil | val_primary/surface_pressure_mae | 22.537 | #2924 |
| AirfRANS | val_primary/surface_mse | 0.000482 | #2951 |
| DrivAerML | val_primary/surface_rel_l2_pct | 3.997% | #2898 |

## Expected outcome

Wall-distance is a direct physical signal for boundary layer behavior. It should help most on AirfRANS and TandemFoil where wake/surface pressure gradients dominate error. DrivAerML will be an interesting test — if the 3D geometry is already well-captured by Fourier coordinates, the gain may be smaller.

**Key diagnostic**: Check whether the model error maps correlate with distance from the wall — if they do, this feature should help directly.

## Reporting

Please report:
1. W&B run IDs for all 3 datasets
2. Best val_primary metric for each dataset  
3. Whether implementation required log-normalization or raw wall-distance worked better
4. Error map plots comparing baseline vs this model near-wall vs far-field (if available)

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT emma:

## Results

### Summary

Wall distance as an explicit physics-informed input feature **does not improve any of the three benchmarks** at the current epoch counts. One dataset (AirfRANS) diverged catastrophically after a cosine warm restart, and DrivAerML in surface-only mode receives a constant-zero wall distance (effectively a control experiment).

### Metrics

| Dataset | Metric | Baseline | This PR (best) | Epoch | Delta |
|---------|--------|----------|---------------|-------|-------|
| TandemFoil | val_primary/surface_pressure_mae | **22.537** | 26.580 | 82 | +18.0% ❌ |
| AirfRANS | val_primary/surface_mse | **0.000482** | 0.00197 | 118 | +4.1× ❌ |
| DrivAerML | val_primary/surface_rel_l2_pct | **3.997%** | 5.276% | 180 | +32.0% ❌ |

⚠️ TandemFoil (ep82) and DrivAerML (ep180) are still running at time of report. Convergence rate suggests neither will reach baseline — see W&B for live tracking.

### W&B Run IDs

- **TandemFoil**: `8zsc67kx` — still running
- **AirfRANS**: `riz9w71y` — killed at ep232 (NaN since ep174)
- **DrivAerML**: `1uoegjee` — still running

All runs in group `sdf-wall-distance-cross-dataset`.

### Implementation Details

- **Wall distance transform**: `log(1 + dist / mean_dist)` — log-normalization was used from the start as recommended in the PR. Raw wall distance was not tested separately.
- **TandemFoil path**: Wall distance integrated into `TandemTargetTransform.prepare_batch`, computed via chunked `cdist` from all points to surface points. Surface points get small but non-zero distance to nearest other surface point.
- **AirfRANS/DrivAerML path**: Wall distance appended at batch time via `augment_batch_wall_distance`. Surface points explicitly get `wall_dist=0`, volume points get actual nearest-surface distance.
- **Input dim**: `surface_input_dim += 1` and `volume_input_dim += 1` when enabled.

### Commands Used

```bash
# TandemFoil (GPU 0)
CUDA_VISIBLE_DEVICES=0 python train.py --dataset tandemfoil --optimizer lion --lr 1.25e-4 --cosine-t-max 10 --grad-clip 0.5 --weight-decay 1e-2 --model-slices 64 --model-layers 3 --model-hidden-dim 192 --model-heads 3 --enable-fourier --enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction --enable-pressure-prior-addition --epochs 999 --ema-decay 0.999 --enable-wall-distance --agent emma --wandb-name "emma/sdf-wall-distance-tandemfoil" --wandb-group sdf-wall-distance-cross-dataset

# AirfRANS (GPU 2)
CUDA_VISIBLE_DEVICES=2 python train.py --dataset airfrans --airfrans-task full --optimizer adamw --lr 6e-4 --cosine-t-max 50 --grad-clip 1.0 --weight-decay 1e-2 --no-use-ema --enable-fourier --model-layers 2 --model-hidden-dim 256 --model-heads 4 --epochs 999 --enable-wall-distance --agent emma --wandb-name "emma/sdf-wall-distance-airfrans" --wandb-group sdf-wall-distance-cross-dataset

# DrivAerML (GPU 4)
SENPAI_MAX_EPOCHS=9999 CUDA_VISIBLE_DEVICES=4 python train.py --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 --model-heads 8 --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200 --enable-wall-distance --agent emma --wandb-name "emma/sdf-wall-distance-drivaerml" --wandb-group sdf-wall-distance-cross-dataset
```

### Peak Memory

- TandemFoil (GPU 0): ~44 GB / 98 GB
- AirfRANS (GPU 2): ~36 GB / 98 GB
- DrivAerML (GPU 4): ~6 GB / 98 GB

### What Happened

**TandemFoil**: The wall distance feature is being computed correctly (chunked cdist, log-normalized), and training converges normally. However, the extra input feature appears to add noise rather than signal — the model is converging 18% above baseline even after 82 epochs of cosine annealing (T_max=10). The improvement rate is ~0.4 MAE per cosine trough cycle, decelerating. It would need ~30+ more cycles (300+ epochs) to hypothetically reach baseline, and convergence rate suggests it will plateau well above.

**AirfRANS** (Critical finding): Training diverged catastrophically. Best metric was 0.00197 at ep118, already 4.1× worse than baseline. After the cosine warm restart near ep150 (T_max=50), val_primary/surface_mse spiked from ~0.005 → 0.045 (ep156) → 0.443 (ep163) → NaN (ep174). The model never recovered. **Important note**: AirfRANS already contains an SDF feature as column 4 of its input data. The wall distance we added is therefore partially redundant — the divergence may be caused by the additional feature altering the input distribution enough to destabilize the optimizer during high-LR restarts. Reducing LR or using EMA might help, but the fundamental signal is already present in the existing features.

**DrivAerML** (Design limitation): DrivAerML runs in surface-only mode (`--drivaerml-train-surface-points 50000`), meaning all query points are on the car surface. Surface points receive `wall_dist=0` by construction. The model is therefore receiving an extra constant-zero column — this is effectively a **control experiment**. The 32% gap vs baseline is likely due to the extra parameter overhead from the wider input layer, not the feature content. The model is still slowly converging (best 5.276% at ep180 vs baseline 3.997%).

### Key Takeaways

1. **Wall distance adds noise, not signal** for these architectures — the Fourier embeddings and existing physics features (Cp panel, TE coord frame, etc.) already capture wall proximity implicitly.
2. **AirfRANS already has SDF** in its input features — adding it again is redundant and causes instability.
3. **DrivAerML surface-only mode makes wall distance meaningless** — all points are on the wall.
4. **Cosine warm restart + extra features = instability risk** — the AirfRANS divergence pattern suggests that additional low-signal features can destabilize training during LR restarts.

### Suggested Follow-ups

1. **If pursuing wall distance**: Test on AirfRANS volume-only or full-field evaluation where wall distance has actual variation, and with EMA + lower peak LR to prevent divergence.
2. **Better geometric features**: Instead of raw wall distance, consider local curvature, surface normals, or boundary layer thickness estimation — these encode richer physics than a scalar distance.
3. **Feature selection investigation**: Run ablation on the existing TandemFoil features to identify which physics features contribute most vs add noise.
4. **For DrivAerML**: Wall distance would only be informative if the model predicted volume fields. For surface-only mode, a local surface curvature feature would be more appropriate.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #3023 because Closing as confirmed negative result. All three datasets regressed significantly:

- TandemFoil: 26.580 vs 22.537 baseline (+18%)
- AirfRANS: NaN/diverged, best 0.00197 vs 0.000482 baseline (+4.1x), killed at ep232
- DrivAerML: 5.276% vs 3.997% baseline (+32%)

The AirfRANS run crashed with NaN at epoch 174 after a cosine warm restart, suggesting the wall-distance feature disrupts training stability. For DrivAerML, surface-only mode receives constant-zero wall distance, making it an ineffective control.

The hypothesis is physically motivated — wall distance IS important in RANS turbulence models — but log-normalized cdist-based wall distance did not provide useful information beyond what Fourier position embeddings already capture. The extra feature may actually hurt optimization by adding a noisy or degenerate signal. Logging as confirmed negative result..
